packages/integration-sdk-runtime/src/storage/memory.ts

```diff
@@ -0,0 +1,239 @@
+      }
+    }
+  }
+
+  addRelationships(stepId: string, newRelationships: Relationship[]): void {
+    for (const relationship of newRelationships) {
+      this.relationshipKeyToRelationshipMap.set(relationship._key, {
+        stepId,
+        relationship,
+      });
+
+      const relationshipTypeKeysMap = this.relationshipTypeToKeysMap.get(
+        relationship._type,
+      );
+
+      if (relationshipTypeKeysMap) {
+        relationshipTypeKeysMap.set(relationship._key, true);
+      } else {
+        this.relationshipTypeToKeysMap.set(
+          relationship._type,
+          new Map<string, boolean>([[relationship._key, true]]),
+        );
+      }
+    }
+  }
+
+  findEntity(_key: string): Entity | undefined {
+    return this.entityKeyToEntityMap.get(_key)?.entity;
+  }
+
+  async iterateEntities<T extends Entity = Entity>(
+    filter: GraphObjectFilter,
+    iteratee: GraphObjectIteratee<T>,
+  ): Promise<void> {
+    const entityTypeKeysMap = this.entityTypeToKeysMap.get(filter._type);
+
+    if (!entityTypeKeysMap) {
+      return;
+    }
+
+    for (const [_key] of entityTypeKeysMap) {
+      const graphObjectData = this.entityKeyToEntityMap.get(_key);
+
+      if (!graphObjectData) {
+        // NOTE: This should never happen. Our data structures should stay in
+        // sync.
+        throw new IntegrationMissingKeyError(
+          `Failed to find entity (_type=${filter._type}, _key=${_key})`,
+        );
+      }
+
+      await iteratee(graphObjectData.entity as T);
+    }
+  }
+
+  async iterateRelationships<T extends Relationship = Relationship>(
+    filter: GraphObjectFilter,
+    iteratee: GraphObjectIteratee<T>,
+  ): Promise<void> {
+    const relationshipTypeKeysMap = this.relationshipTypeToKeysMap.get(
+      filter._type,
+    );
+
+    if (!relationshipTypeKeysMap) {
+      return;
+    }
+
+    for (const [_key] of relationshipTypeKeysMap) {
+      const graphObjectData = this.relationshipKeyToRelationshipMap.get(_key);
+
+      if (!graphObjectData) {
+        // NOTE: This should never happen. Our data structures should stay in
+        // sync.
+        throw new IntegrationMissingKeyError(
+          `Failed to find relationship (_type=${filter._type}, _key=${_key})`,
+        );
+      }
+
+      await iteratee(graphObjectData.relationship as T);
+    }
+  }
+
+  /**
+   * Instead of clearing the entire map, we want to delete old data individually.
+   * This will allow us to eventually run steps in parallel without locking when
+   * steps have no interdependence. If we cleared it out, it's possible that
+   * two steps could be running in parallel and one of the steps may clear out
+   * the maps while the other is still relying on it.
+   */
+  flushEntities(entities: Entity[]) {
+    for (const entity of entities) {
+      this.entityKeyToEntityMap.delete(entity._key);
+      const entityTypeKeysMap = this.entityTypeToKeysMap.get(entity._type);
+      entityTypeKeysMap?.delete(entity._key);
```

Comment:
I wonder if we should really conditionally chain `entityTypeKeysMap?.delete(entity._key)` here. If `entityTypeKeysMap` is undefined, doesn't this represent an out-of-sync memory data store, something we would consider an exception?

Reply:
Yeah, it seems like that kind of situation would be an indicator of syncing issues. Throwing an exception sounds reasonable to me, but if you end up thinking that that the absence of the key map shouldn't cause a step to fail, perhaps consider logging out a warning or error message.

Reply:
Agreed. This would be indicative of a bug. I'm changing it to throw.

Reply:
Fixed